v3/examples/mutual_tls_echo/main.go

```diff
@@ -17,104 +17,89 @@ import (
+		defer pingTimer.Stop()
+		for {
+			pingCtx, pingCancel := context.WithTimeout(ctx, operationTimeout)
+			if err := client.PingAndWait(pingCtx); err != nil {
+				fmt.Printf("[TLS Echo] Failed ping %v\n", err)
+			}
+			pingCancel()
+			select {
+			case <-pingCtx.Done():
+			case <-pingTimer.C:
+			case <-ctx.Done():
+				return
 			}
-		case <-ctx.Done():
-			fmt.Println("ping timeout")
-			gotDisconnected <- true
-		}
 
-		fmt.Println("Ping sent and ack'ed")
-		// Wait for the timer to before we loop again
-		<-ticker.C
-	}
+		}
+	}()
 }
 
-func waitForAck(delegate *mode.ModeMqttDelegate) uint16 {
-	// Wait for the ack, or timeout
-	ctx, cancel := context.WithTimeout(context.Background(),
-		time.Second)
-	defer cancel()
-
+func waitForAck(ctx context.Context, delegate *mode.ModeMqttDelegate) uint16 {
 	// Block on the return channel or timeout
 	select {
 	case queueResp := <-delegate.QueueAckCh:
 		if queueResp.Err != nil {
-			fmt.Printf("Queued request failed: %s\n", queueResp.Err)
+			fmt.Printf("[TLS Echo] Queued request failed: %s\n", queueResp.Err)
 		} else {
 			return queueResp.PacketID
 		}
 	case <-ctx.Done():
-		fmt.Printf("Ack response timeout: %s\n", ctx.Err())
+		fmt.Printf("[TLS Echo] Ack response timeout: %s\n", ctx.Err())
 	}
 
 	return 0
 }
 
-func receiveCommands(cmdChannel chan *mode.DeviceCommand,
-	delegate *mode.ModeMqttDelegate, client *mode.MqttClient) {
-
-	for cmd := range cmdChannel {
-		switch cmd.Action {
-		case "doEcho":
-			var params struct {
-				Msg string `json:"msg"`
-			}
-
-			if err := cmd.BindParameters(&params); err != nil {
-				fmt.Printf("Failed to bind command parameters: %s\n", err.Error())
-				return
-			}
-
-			event := mode.DeviceEvent{
-				EventType: "echo",
-				EventData: map[string]interface{}{"msg": params.Msg},
-				Qos:       mode.QOSAtLeastOnce,
-			}
-
-			_, err := client.PublishEvent(event)
-			if err != nil {
-				fmt.Printf("Failed to send event: %s\n", err.Error())
+func receiveCommands(ctx context.Context, delegate *mode.ModeMqttDelegate, client *mode.MqttClient) {
+	cmdChannel := delegate.GetCommandChannel()
+	for {
+		select {
+		case cmd := <-cmdChannel:
+			fmt.Printf("[TLS Echo][Echo] Received command: %s\n", cmd.Action)
+			switch cmd.Action {
+			case "doEcho":
+				var params struct {
+					Msg string `json:"msg"`
+				}
+
+				if err := cmd.BindParameters(&params); err != nil {
+					fmt.Printf("[TLS Echo] Failed to bind command parameters: %s\n", err.Error())
+					return
+				}
+
+				event := mode.DeviceEvent{
+					EventType: "echo",
+					EventData: map[string]interface{}{"msg": params.Msg},
+					Qos:       mode.QOSAtLeastOnce,
+				}
+				ctx, cancel := context.WithTimeout(context.Background(), operationTimeout)
+				defer cancel()
+				_, err := client.PublishEvent(ctx, event)
+				if err != nil {
+					fmt.Printf("[TLS Echo] Failed to send event: %s\n", err.Error())
+				}
+				waitForAck(ctx, delegate)
```

Comment:
This line is in the loop. So, I think this `defer` works only when exit function.
```suggestion
				ctx, cancel := context.WithTimeout(context.Background(), operationTimeout)
				_, err := client.PublishEvent(ctx, event)
				if err != nil {
					fmt.Printf("[TLS Echo] Failed to send event: %s\n", err.Error())
				}
				waitForAck(ctx, delegate)
				cancel()
```

And... It may needs to check result of `ctx.Error()` whether it was returned `context.DeadlineExceeded`.

Reply:
Ah, thanks for catching that. I've also renamed the variable because it's the same as the parameter.